bridge/src/helpers.rs

```diff
@@ -0,0 +1,231 @@
+// Copyright 2017 Parity Technologies (UK) Ltd.
+// This file is part of Parity-Bridge.
+
+// Parity-Bridge is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity-Bridge is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity-Bridge.  If not, see <http://www.gnu.org/licenses/>.
+
+//! various helper functions
+
+use error::{self, ResultExt};
+use ethabi::{self, ContractFunction, ParseLog, RawLog};
+use futures::future::FromErr;
+use futures::{Async, Future, Poll, Stream};
+use serde::de::Error;
+use serde::{Deserialize, Deserializer, Serializer};
+use std::time::Duration;
+use tokio_timer::{Timeout, Timer};
+use web3::api::Namespace;
+use web3::helpers::CallResult;
+use web3::types::{Address, Bytes, CallRequest, H256, TransactionRequest, U256};
+use web3::{self, Transport};
+
+/// attempts to convert a raw `web3_log` into the ethabi log type of a specific `event`
+pub fn parse_log<T: ParseLog>(event: &T, web3_log: &web3::types::Log) -> ethabi::Result<T::Log> {
+    let ethabi_log = RawLog {
+        topics: web3_log.topics.iter().map(|t| t.0.into()).collect(),
+        data: web3_log.data.0.clone(),
+    };
+    event.parse_log(ethabi_log)
+}
+
+/// use `AsyncCall::new(transport, contract_address, timeout, function)` to
+/// get a `Future` that resolves with the decoded output from calling `function`
+/// on `contract_address`.
+pub struct AsyncCall<T: Transport, F: ContractFunction> {
+    future: Timeout<FromErr<CallResult<Bytes, T::Out>, error::Error>>,
+    function: F,
+}
+
+impl<T: Transport, F: ContractFunction> AsyncCall<T, F> {
+    /// call `function` at `contract_address`.
+    /// returns a `Future` that resolves with the decoded output of `function`.
+    pub fn new(transport: &T, contract_address: Address, timeout: Duration, function: F) -> Self {
+        let payload = function.encoded();
+        let request = CallRequest {
+            from: None,
+            to: contract_address,
+            gas: None,
+            gas_price: None,
+            value: None,
+            data: Some(Bytes(payload)),
+        };
+        let inner_future = web3::api::Eth::new(transport)
+            .call(request, None)
+            .from_err();
```

Comment:
it converts the futures contained error (type). https://docs.rs/futures/0.1/futures/future/trait.Future.html#method.from_err. required here because the error type of the `CallResult` returned by `.call(request, None)` is `web3::error::Error` which makes `CallResult` a future that can't be passed directly to `timer.timeout` since `timer.timeout` requires the error to satisfy `From<TimeoutError<F>>`. thus `from_err` is used to convert to `bridge::Error` which satisfies that constraint https://github.com/paritytech/parity-bridge/pull/159/files/07d78782fbffbf7acd2e3a4e2df455df467ca960#diff-7fff022d65a8e238f9696f70844d5831R70